### PR TITLE
Create UserModel class

### PR DESCRIPTION
**Pull request by NoB0**
This is a first attempt at modelling a user. The model stores the previous choices (e.g., accept, reject, and dont_like). The preference towards a set of movies can be estimated. 

Note that the tag preferences are not computationally efficient as of now (e.g., a database call and exploration of the results are performed each time to get the tag preference).

### CODE REVIEW

**Comment by IKostric** [comment]
The model right now estimates the preferances for slots from liked/disliked movies. We should also keep track of explicitly stated preferances, e.g., "I like drama".

File: moviebot/user_modeling/user_model.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        The JSON file with historical movie choices is a dictionary with the
+        movie id as key and the list of choices as value.
+
+        Args:
+            user_id: User id.
+            historical_movie_choices_path: Path to the JSON file with historical
+              movie choices. Defaults to None.
+        """
+        self.user_id = user_id
+        self._movies_choices = defaultdict(str)
```
Comment (IKostric) [inline moviebot/user_modeling/user_model.py:25]:
```suggestion
        self._movies_choices = defaultdict(list)
```

File: moviebot/user_modeling/user_model.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+        if historical_movie_choices_path and os.path.exists(
+            historical_movie_choices_path
+        ):
+            # Load historical movie choices
+            movie_choices = json.load(historical_movie_choices_path)
```
Comment (IKostric) [inline moviebot/user_modeling/user_model.py:31]: I believe `json.load`` expects a file object, not path.

File: moviebot/user_modeling/user_model.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        sql_cursor = database.sql_connection.cursor()
+        tag_set = sql_cursor.execute(
+            f"SELECT ID FROM {database._get_table_name()} WHERE {slot} LIKE "
+            f"'%{tag}%'"
+        ).fetchall()
+
+        preference = 0.0
+        count_rated = 0
+        for movie_id, choices in self._movies_choices.items():
+            if movie_id in tag_set:
```
Comment (IKostric) [inline moviebot/user_modeling/user_model.py:115]: I believe `fetchall` returns a list of tuples. You might want to extract first element of the tuple to get the list of ids.
Additionally, you might want to convert the list to `set()` to get a much faster lookup.

```
tag_set = {row[0] for row in tag_set}
```

File: moviebot/user_modeling/user_model.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        return self._movies_choices
+
+    def update_movie_choice(self, movie_id: str, choice: str) -> None:
+        """Updates the choices for a given user.
+
+        Args:
+            movie_id: Id of the movie.
+            choice: User choice (i.e., accept, reject).
+        """
+        self._movies_choices[movie_id] = self._movies_choices[movie_id] + [
```
Comment (IKostric) [inline moviebot/user_modeling/user_model.py:48]: Should the latest choice be appended to or overwrite the previous preference?

File: moviebot/user_modeling/user_model.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            f"'%{tag}%'"
+        ).fetchall()
+
+        preference = 0.0
+        count_rated = 0
+        for movie_id, choices in self._movies_choices.items():
+            if movie_id in tag_set:
+                # TODO: decide how to handle contradictory choices (e.g., the
+                # same movie was accepted and rejected)
+                for choice in choices:
```
Comment (IKostric) [inline moviebot/user_modeling/user_model.py:118]: Why do we have multiple choices for each movie? If a user changes their mind, shouldn't we simply use the latest preference?

File: moviebot/user_modeling/user_model.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        self._movies_choices = defaultdict(str)
+
+        if historical_movie_choices_path and os.path.exists(
+            historical_movie_choices_path
+        ):
+            # Load historical movie choices
+            movie_choices = json.load(historical_movie_choices_path)
+            self._movies_choices.update(movie_choices)
+
+        self.tag_preferences = defaultdict(defaultdict(float))
```
Comment (IKostric) [inline moviebot/user_modeling/user_model.py:34]: Default dict needs a callable as first argument.
```suggestion
        self.tag_preferences = defaultdict(lambda: defaultdict(float))
```

**Review by IKostric — CHANGES_REQUESTED** [review]
See inline comments. I have several questions regarding choices. The main one being: Why keep historical choices instead of the latest one only?

File: moviebot/user_modeling/user_model.py (hunk 6 above)
Comment (NoB0) [reply moviebot/user_modeling/user_model.py:34]: Done.

**Comment by NoB0** [comment]
> See inline comments. I have several questions regarding choices. The main one being: Why keep historical choices instead of the latest one only?

Choices include various values for different intents, for example, the choice can `watched`, `dont_like` when a recommendation is rejected. These may be handled differently when converting the choices to preferences.
As of now, there is no difference between historical choices and session choices and the tag preference is inferred from all the choices available. This allows us to infer tag preference for tag that are not mentioned in the current conversation. 

Based on your comments, I would make the following suggestion:
  - Use historical tag preferences instead of historical movie choice in `__init__`
  - Store movie choices only for the current session and use them to compute unknown tag preference
Would this make the code and logic more clear?

**Review by kbalog — CHANGES_REQUESTED** [review]
It would be good to get clarity on the purpose of this class. To me, it serves two main roles:
1) provide a structured representation of the preferences the user expressed to the recommender model; this could be in the form of tags with weights or could be just a list of NL utterances that expressed a preference; it depends on the type of recommender model which format they can take as input
2) provide input to the explainability component that would generate an explainable user model. As with 1) this could be either in the form of raw/annotated NL utterances or tags with weights.

Given the above, this class should have only two public methods: `get_utterances_with_preference()` and `get_tags_preferences()`. All other methods should be private.

**Comment by kbalog** [comment]
We might want to distinguish between item and slot preferences... Not sure if tags are represented as a slot.

File: moviebot/user_modeling/user_model.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
+"""Class for user modeling.
+
+The user model stores the user's preferences, in terms of slots and items in a
+structured (dictionary) and unstructured manner (utterances). These preferences
+serves to main purposes:
```
Comment (IKostric) [inline moviebot/user_modeling/user_model.py:5]:
```suggestion
serves two main purposes:
```

**Review by IKostric — APPROVED** [review]
LGTM, This will likely need to be iterated on, but it is better to merge and see how we will end up using it.

**Review by kbalog — CHANGES_REQUESTED** [review]
See my comments. How about tests for this class?

File: moviebot/user_modeling/user_model.py [hunk 8]
```diff
@@ -1,71 +1,137 @@
...
-            choice: User choice (i.e., accept, reject).
+        Returns:
+            User model.
         """
-        self._movies_choices[movie_id].append(choice)
-
-    def update_movies_choices(self, movies_choices: Dict[str, str]) -> None:
-        """Updates the movie choices for a given user.
+        user_model = cls()
+        if os.path.exists(json_path):
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/user_modeling/user_model.py:55]: Good point. Done.

File: moviebot/user_modeling/user_model.py [hunk 9]
```diff
@@ -1,71 +1,137 @@
...
+            "intent": utterance.intent.label,
+            "slot_values": [
+                [annotation.slot, annotation.value]
+                for annotation in utterance.annotations
+            ]
+            if utterance.annotations
+            else [],
+        }
+
+    def save(self, json_path: str) -> None:
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/user_modeling/user_model.py:105]: Done.

File: moviebot/user_modeling/user_model.py [hunk 10]
```diff
@@ -1,71 +1,137 @@
...
-    def update_movies_choices(self, movies_choices: Dict[str, str]) -> None:
-        """Updates the movie choices for a given user.
+        user_model = cls()
+        if os.path.exists(json_path):
+            user_model_json = json.load(open(json_path, "r"))
+            user_model.slot_preferences.update(
+                user_model_json["slot_preferences"]
+            )
+            for slot, utterance in user_model_json[
+                "slot_preferences_nl"
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/user_modeling/user_model.py:61]: Done.

File: moviebot/user_modeling/user_model.py [hunk 11]
```diff
@@ -1,71 +1,137 @@
...
+            Dict[str, AnnotatedUtterance]
+        ] = defaultdict(lambda: defaultdict(list))
+
+        # Structured and unstructured item preferences
+        self.item_preferences: Dict[Dict[str, float]] = defaultdict(
+            lambda: defaultdict(float)
+        )
+        self.item_preferences_nl: Dict[
+            Dict[str, AnnotatedUtterance]
+        ] = defaultdict(lambda: defaultdict(list))
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/user_modeling/user_model.py:39]: I updated the types. The idea is that for slots we have a dict of dicts ({slot:{value1:preference1, value2:preference2}}) and for items we have a single dict ({item: preference}).

File: moviebot/user_modeling/user_model.py [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+        else:
+            raise FileNotFoundError(f"JSON file {json_path} not found.")
+        return user_model
+
+    def _utterance_to_dict(
+        self, utterance: AnnotatedUtterance
+    ) -> Dict[str, str]:
+        """Converts an utterance to a dictionary.
+
+        TODO: Move this method to DialogueKit AnnotatedUtterance class.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/user_modeling/user_model.py:85]: Done.

File: moviebot/user_modeling/user_model.py [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+            Preference within the range [-1,1].
+        """
+        if choice == "accept":
+            return 1.0
+        elif choice in ["reject", "dont_like"]:
+            return -1.0
+
+        return 0.0
+
+    def update_item_preference(self, item: str, choice: str) -> None:
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/user_modeling/user_model.py:157]: Good point, this is especially true now that we have different types of NLU. I move this function to `recommendation_decision_processing.py`, and add a new constraint to attach a preference to a dialogue act.

File: moviebot/nlu/user_intents_checker.py [hunk 14]
```diff
@@ -334,15 +338,27 @@ def check_reject_intent(
...
             ]
         ):
             dact.intent = UserIntents.REJECT
-            dact.params = [ItemConstraint("reason", Operator.EQ, "watched")]
+            preference = convert_choice_to_preference(
+                RecommendationChoices.WATCHED
+            )
+            dact.params = [
+                ItemConstraint("reason", Operator.EQ, "watched"),
+                ItemConstraint("preference", Operator.EQ, preference),
```
Comment (kbalog) [inline moviebot/nlu/user_intents_checker.py:360]: Please create an issue for creating enums for the possible contraints that can be expressed ("reason", "perference", etc.)

File: moviebot/user_modeling/user_model.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kbalog) [reply moviebot/user_modeling/user_model.py]: Please make sure to also mention this in the comments above the variable declarations.

**Review by kbalog — APPROVED** [review]
Few comments remain, but LGTM-ed from my side.

File: moviebot/user_modeling/user_model.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/user_modeling/user_model.py]: Done.

File: moviebot/nlu/recommendation_decision_processing.py
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/nlu/recommendation_decision_processing.py:1]: Follow-up issue #226

File: moviebot/nlu/user_intents_checker.py (hunk 14 above)
Comment (NoB0) [reply moviebot/nlu/user_intents_checker.py:360]: Follow-up issue #225

**Pull request merged**